lib/arm/cpu_features.c

```diff
@@ -111,9 +113,43 @@ void setup_cpu_features(void)
 		features |= ARM_CPU_FEATURE_NEON;
 	if (hwcap & (1 << 4))	/* HWCAP_PMULL */
 		features |= ARM_CPU_FEATURE_PMULL;
+	if (hwcap & (1 << 7))	/* HWCAP_CRC32 */
+		features |= ARM_CPU_FEATURE_CRC32;
 #endif
 
 	_cpu_features = features | ARM_CPU_FEATURES_KNOWN;
 }
 
+#elif defined(__aarch64__) /* not __linux__ */
+
+volatile u32 _cpu_features = 0;
+
+void setup_cpu_features(void)
+{
+	/* NEON is a mandatory part of AArch64 */
+	_cpu_features = ARM_CPU_FEATURE_NEON;
+
+	/* On FreeBSD >= 12, Linux >= 4.11 and potentially others, it is possible to use
+	   privileged system registers from userspace to check CPU feature support.
+
+	   For proper support of SoCs where different cores have different capabilities,
+	   the OS has to always report only the features supported by all cores, like FreeBSD does. */
+
+#define bits_shift(x, high, low) ((x >> low) & ((1 << (high - low + 1)) - 1))
+
+#if __FreeBSD__ >= 12
+	uint64_t isar0 = 0;
+	__asm__("mrs %0, ID_AA64ISAR0_EL1" : "=r"(isar0));
+
+	if (bits_shift(isar0, 7, 4) >= 1)
+		_cpu_features |= ARM_CPU_FEATURE_PMULL;
+	if (bits_shift(isar0, 19, 16) >= 1)
```

Comment:
[AArch64 documentation](https://developer.arm.com/docs/ddi0595/latest/aarch64-system-registers/id_aa64isar0_el1) says this bitfield contains a value, not individual bits, and only 0 and 1 are defined currently.  So I think this should be `== 1`, not `>= 1`.

Reply:
Yes, technically "From Armv8.1, the only permitted value is 0b0001", but all these bitfields work like increasing levels of support (look at AES and SHA2).